crypto-pgpainless/src/main/kotlin/dev/msfjarvis/aps/crypto/PGPKeyManager.kt

```diff
@@ -50,63 +63,105 @@ public class PGPKeyManager(
+
+        val keys = keyFiles.map { file -> Key(file.readBytes()) }
+        // Try to parse the key ID as an email
+        val selector = SelectUserId.byEmail(id)
+        val userIdMatch =
+          keys.map { key -> key to tryParseKeyring(key) }.firstOrNull { (_, keyRing) ->
+            selector.firstMatch(keyRing) != null
+          }
+
+        if (userIdMatch != null) {
+          return@runCatching userIdMatch.first
+        }
+
+        val keyIdMatch =
+          keys.map { key -> key to tryGetId(key) }.firstOrNull { (_, keyId) ->
+            keyId == id || keyId == "0x$id"
+          }
 
-        for (keyFile in keys) {
-          val secretKeyRing = PGPainless.readKeyRing().secretKeyRing(keyFile.inputStream())
-          val secretKey = secretKeyRing.secretKey
-          val keyPair = PGPKeyPair(secretKey)
-          if (keyPair.getKeyId() == id) return@runCatching keyPair
+        if (keyIdMatch != null) {
+          return@runCatching keyIdMatch.first
         }
 
         throw KeyManagerException.KeyNotFoundException(id)
       }
     }
 
-  override suspend fun getAllKeys(): Result<List<PGPKeyPair>, Throwable> =
+  override suspend fun getAllKeys(): Result<List<Key>, Throwable> =
     withContext(dispatcher) {
       runCatching {
         if (!keyDirExists()) throw KeyManagerException.KeyDirectoryUnavailableException
         val keys = keyDir.listFiles()
         if (keys.isNullOrEmpty()) return@runCatching listOf()
-
-        keys
-          .map { keyFile ->
-            val secretKeyRing = PGPainless.readKeyRing().secretKeyRing(keyFile.inputStream())
-            val secretKey = secretKeyRing.secretKey
-
-            PGPKeyPair(secretKey)
-          }
-          .toList()
+        keys.map { keyFile -> Key(keyFile.readBytes()) }.toList()
       }
     }
 
+  override suspend fun getKeyId(key: Key): String? = tryGetId(key)
+
+  // TODO: This is a temp hack for now and in future it should check that the GPGKeyManager can
+  // decrypt the file.
   override fun canHandle(fileName: String): Boolean {
-    // TODO: This is a temp hack for now and in future it should check that the GPGKeyManager can
-    // decrypt the file
     return fileName.endsWith(KEY_EXTENSION)
   }
 
+  /** Checks if [keyDir] exists and attempts to create it if not. */
   private fun keyDirExists(): Boolean {
     return keyDir.exists() || keyDir.mkdirs()
   }
 
+  /**
+   * Attempts to parse a [PGPKeyRing] from a given [key]. The key is first tried as a secret key and
+   * then as a public one before the method gives up and returns null.
+   */
+  private fun tryParseKeyring(key: Key): PGPKeyRing? {
+    val secKeyRing = runCatching { PGPainless.readKeyRing().secretKeyRing(key.contents) }.get()
+    if (secKeyRing != null) {
+      return secKeyRing
+    }
+    val pubKeyRing = runCatching { PGPainless.readKeyRing().publicKeyRing(key.contents) }.get()
+    if (pubKeyRing != null) {
+      return pubKeyRing
+    }
+    return null
```

Comment:
Could be simplified to `return pubKeyRing`

Reply:
I prefer the consistency with `secKeyRing`